streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java

```diff
@@ -0,0 +1,792 @@
+                            // it's important that this step happens before updating the current
+                            // segment. if there is a partial failure (this step succeeds but the
+                            // update to the current segment fails), updating the older segment
+                            // first means there will not be data loss. (rather, there will be
+                            // duplicated data which is fine as it can/will be reconciled later.)
+                            final T olderSegment = versionedStoreClient
+                                .getOrCreateSegmentIfLive(segmentIdForTimestamp, context, observedStreamTime);
+                            if (olderSegment != null) {
+                                final byte[] olderSegmentValue = olderSegment.get(key);
+                                if (olderSegmentValue == null) {
+                                    olderSegment.put(
+                                        key,
+                                        segmentValueSchema.newSegmentValueWithRecord(
+                                            searchResult.value(), searchResult.validFrom(), timestamp
+                                        ).serialize()
+                                    );
+                                } else {
+                                    final SegmentValue olderSv = segmentValueSchema.deserialize(olderSegmentValue);
+                                    olderSv.insertAsLatest(searchResult.validFrom(), timestamp, searchResult.value());
+                                    olderSegment.put(key, olderSv.serialize());
+                                }
+                            }
+
+                            // update in newer segment (replace the record that was just moved with the new one)
+                            sv.updateRecord(timestamp, value, searchResult.index());
+                            segment.put(key, sv.serialize());
+                        } else {
+                            sv.insert(timestamp, value, searchResult.index());
+                            segment.put(key, sv.serialize());
+                        }
+                    }
+                    return new PutStatus(true, foundTs);
+                }
+
+                if (foundMinTs < observedStreamTime - historyRetention) {
+                    // the record being inserted does not affect version history. discard and return
+                    if (expiredRecordSensor.isPresent()) {
+                        expiredRecordSensor.get().record(1.0d, ProcessorContextUtils.currentSystemTime(context));
+                        LOG.warn("Skipping record for expired put.");
+                    }
+                    return new PutStatus(true, foundTs);
+                }
+
+                // it's possible the record belongs in this segment, but also possible it belongs
+                // in an earlier segment. mark as tentative and continue. as an optimization, this
```

Comment:
"earlier segment" -> or "latest value store", right?

Reply:
No, this case is only hit if we've already found a segment which contains data later than the current insertion timestamp, so the new record will never be put into the latest value store.